Add kotlin.serialization implementation of Serializer

lion7 opened the pull request:
This PR adds support for serializing by using `kotlinx.serialization` and is basically a continuation of https://github.com/AxonFramework/extension-kotlin/pull/124.

Note that I had to upgrade the Kotlin version (and fix a few generics in `QueryGatewayExtensions.kt`) to make it work.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AxonFramework/extension-kotlin?pullRequest=338) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AxonFramework/extension-kotlin?pullRequest=338) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AxonFramework/extension-kotlin?pullRequest=338) it.</sub>

jangalinski commented:
Great approach. We are working on a kotlinx.serialization based avro serializer as well and might (re-)use generic features provided by extension-kotlin.
But: kotlinx.serialization is a generic tool for serialization, json is just one aspect. So I would like to have a clearer separation between json-specific and kotlinx-generic features ...

lion7 commented:
This implementation is already independent of the serialization format. I changed the main dependency to use the core library instead and added a few more tests for binary formats as well. 

It should work with any `SerialFormat` that either implements `StringFormat` or `BinaryFormat`.

gklijs commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,333 @@
+        )
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
```
It would be nice if we could refrain from using experimental api's in src. Although small, there is a risk these will change and not be compatible with future versions of Kotlin.

lion7 replied:
This was actually no longer needed, so I removed the opt-in

gklijs commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,333 @@
+    contextual(QuartzScheduleToken::class) { QuartzScheduleTokenSerializer }
+    polymorphic(ScheduleToken::class) {
+        subclass(SimpleScheduleTokenSerializer)
+        subclass(QuartzScheduleTokenSerializer)
```
Does this mean when developers use JobRunr or DbScheduler for deadlines/sheduled events they should add something of their own?

lion7 replied:
Yes, they would have to write their own serializer and add it to a `SerializersModule`.
Note that you can combine multiple `SerializerModule`s into a single one that contains all needed serializers. 

The main reason for this is because `kotlinx.serialization` does not use any runtime mechanic such as reflection for serializing classes. Instead the Kotlin compiler generates dedicated serializers for each (data) class annotated with `@Serializable`. For non-Kotlin or external classes this means that you have to write a dedicated serializer and explicitly add it to a `SerializersModule`. You can then add this module to a `SerialFormat` such as `Json` or `Cbor`.

gklijs commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,333 @@
+    contextual(ReplayToken::class) { ReplayTokenSerializer }
+    contextual(GlobalSequenceTrackingToken::class) { GlobalSequenceTrackingTokenSerializer }
+    polymorphic(TrackingToken::class) {
+        subclass(ConfigTokenSerializer)
```
Does this mean for any other token, for example like with the Mongo or Kafka extension, developers have to provide additional configuration?

lion7 replied:
See the other comment on `ScheduleToken`.

jangalinski replied:
Just a thought here: When we build the serializer for avro, and came to this point, we agreed that a custom serializer is only relevant for messaging and event-storage, so we decided to just let the default axon-serializer do the lifting for all internal types like tokens ... Its java code anyway.

gklijs commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,333 @@
+package org.axonframework.extensions.kotlin.serialization
```
I would like to see the copywrite block like:
```
/*
 * Copyright (c) 2010-2021. Axon Framework
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
On all the new files.
```

lion7 replied:
Done

gklijs reviewed (changes requested):
I have some questions about Axon Specific classes that needs to be serialized, which are not included in the `SerializersModule`. All new files should have the copyright notice. It would also be nice if we don't need to opt in for experimental api's.

zambrovski commented:
We were playing with `KSerializer` registration in our Apache Avro approach and came up with a small SPI. By doing so it is pretty easy to write own serializers and let the SPI loader automatically register those. 

But before building this, it would be a good idea to document how new KSerializers needs to be registered. In general, do you have any docs describing the implemented features?

jangalinski commented:
> We were playing with `KSerializer` registration in our Apache Avro approach and came up with a small SPI. By doing so it is pretty easy to write own serializers and let the SPI loader automatically register those.
> 
> But before building this, it would be a good idea to document how new KSerializers needs to be registered. In general, do you have any docs describing the implemented features?

Actually, this part is already public/released ... peek here if you like: https://github.com/toolisticon/avro-kotlin/tree/develop/avro-kotlin-serialization/src/main/kotlin/spi

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/ArrayResponseType.kt:
```diff
@@ -0,0 +1,113 @@
+ * @author Gerard de Leeuw
+ * @see org.axonframework.messaging.responsetypes.MultipleInstancesResponseType
+ */
+class ArrayResponseType<E>(elementType: Class<E>) : AbstractResponseType<Array<E>>(elementType) {
```
@lion7, can you explain to me what the intent is of the `ArrayResponseType` for the `KotlinSerializer` exactly?

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,346 @@
+            val index = decodeElementIndex(descriptor)
+            if (index == CompositeDecoder.DECODE_DONE) break
+            when (index) {
+                0 -> gapIndex = decodeLongElement(descriptor, index)
```
Question, I assume the values 0 and 1 reflecting the fields `index` and `gaps` mean that if we'd reorder the fields in the `GapAwareTrackingToken`, this would break the serialization?

smcvb replied:
@lion7 explained me that the `index` numbers align with the numbers given during the `serialize` invocation. Hence, field reordering has no impact on this process.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,346 @@
+import org.axonframework.messaging.responsetypes.ResponseType
+import kotlin.reflect.KClass
+
+private val trackingTokenSerializer = PolymorphicSerializer(TrackingToken::class).nullable
```
Reading the documentation of the `PolymorphicSerializer` constructor, I read the following:

```txt
...
 * To avoid the most common security pitfalls and reflective lookup (and potential load) of an arbitrary class,
 * all serializable implementations of any polymorphic type must be [registered][SerializersModuleBuilder.polymorphic]
 * in advance in the scope of base polymorphic type, efficiently preventing unbounded polymorphic serialization
 * of an arbitrary type.
...
```

So, I assume that when this `trackingTokenSerializer` is instantiated all implementation have been registered, and novel implementation can be included by users as well?

smcvb replied:
During a chat with @lion7 we concluded that users are capable of adding their own `TrackingToken` serializer when necessary.
If somebody combines Axon Framework with our Mongo and this Kotlin extension, they'd be required to draft a `MongoTrackingTokenSerializer`, after all.
However, to simplify that endeavour, we should make the `trackingTokenSerializer` public.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,346 @@
+    }
+}
+
+object ConfigTokenSerializer : KSerializer<ConfigToken> {
```
Question: Would a user of the Kotlin Extension ever be required to work with the `KSerializer` implementations in this file directly?

smcvb replied:
After a face-to-face discussion with @lion7, we conclude that some could be `private`, and some should be public. 
Since Axon's Kotlin Serialization support has already been published, we can no longer make any of these `private`.

Hence, we'll proceed to add documentation to all these public classes; anything public should be documented, after all.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,197 @@
+ * @since 4.9.1
+ * @author Gerard de Leeuw
+ */
+class KotlinSerializer(
```
Might be nice to have some samples in the class level KDoc on how to construct the `KotlinSerializer`.

smcvb replied:
Pushed some additional documentation directly myself.

smcvb reviewed (commented):
Looks fine, although I have a couple of questions and remarks.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/AxonSerializers.kt:
```diff
@@ -0,0 +1,346 @@
+object ReplayTokenSerializer : KSerializer<ReplayToken> {
+
+    override val descriptor = buildClassSerialDescriptor(ReplayToken::class.java.name) {
+        element<TrackingToken>("tokenAtReset")
```
I have made some local adjustments to validate my assumption here and I think they're correct.

The predicament, however, is the fact that the `context` field is of type `Object`, so `Any` in Kotlin.
After a quick search, my assumption is confirmed that the `Any` type does not work nicely with serialization.

I would wager that if we could set a Serializer for the `context` that would (1) check if the field is a primitive type and (2) check if it's of a type for which the user has provided a Serializer, we should be good.
The former should be a straightforward search, but I am not entirely sure how to solve requirement two for custom types from the user.
So, with that said, do you have a suggestion on how to tackle this, @lion7?

smcvb replied:
After discussion with @lion7 on how to resolve this, we came to the conclusion that any clean solution would require breaking changes in Axon Framework.
Thus, those changes will have to wait until Axon Framework 5.

For the time being, we'll adjust the `ReplayTokenSerializer` to support a `resetContext` of type `String`.
In doing so, a user would only be able to provide `Strings` as the reset context.
Although restrictive, it will allow a user to do the de-/serialization of the reset context themselves before invoking the reset and when handling a reset context in a `@ResetHandler` annotated method.

This difference between base Axon Framework and the Kotlin Extension should, of course, be documented.
This work will be covered in another PR.
